core/src/main/java/tc/oc/pgm/spawner/Spawner.java

```diff
@@ -0,0 +1,148 @@
+      return;
+    }
+    if (match.getTick().tick - lastTick >= generatedDelay) {
+      for (Spawnable spawnable : definition.objects) {
+        Vector randomSpawnLoc = definition.spawnRegion.getRandom(match.getRandom());
+        spawnable.spawn(
+            new Location(
+                match.getWorld(),
+                randomSpawnLoc.getX(),
+                randomSpawnLoc.getY(),
+                randomSpawnLoc.getZ()),
+            match);
+        if (spawnable.isTracked()) {
+          spawnedEntities = spawnedEntities + spawnable.getSpawnCount();
+        }
+      }
+      generateDelay();
+      lastTick = match.getTick().tick;
+    }
+  }
+
+  private void generateDelay() {
+    if (definition.minDelay == definition.maxDelay) {
+      generatedDelay = TimeUtils.toTicks(definition.delay);
+    } else {
+      long maxDelay = TimeUtils.toTicks(definition.maxDelay);
+      long minDelay = TimeUtils.toTicks(definition.minDelay);
+      generatedDelay =
+          (long)
+              (match.getRandom().nextDouble() * (maxDelay - minDelay)
+                  + minDelay); // Picks a random tick duration between minDelay and maxDelay
+    }
+  }
+
+  private boolean canSpawn() {
+    if (spawnedEntities >= definition.maxEntities || trackedPlayers.isEmpty()) {
+      return false;
+    }
+    for (MatchPlayer p : trackedPlayers.values()) {
+      if (definition.playerFilter.query(p.getQuery()).isAllowed()) {
+        return true;
+      }
+    }
+    return false;
+  }
+
+  private boolean isTrackedEntity(Entity entity) {
+    return entity.getMetadata(METADATA_KEY, PGM.get()) != null;
+  }
+
+  @EventHandler(priority = EventPriority.MONITOR)
+  public void onEntityDeath(EntityDeathEvent event) {
+    if (isTrackedEntity(event.getEntity())) {
+      spawnedEntities--;
+    }
+  }
+
+  @EventHandler(priority = EventPriority.MONITOR)
+  public void onItemDespawn(ItemDespawnEvent event) {
+    if (isTrackedEntity(event.getEntity())) {
+      spawnedEntities = spawnedEntities - event.getEntity().getItemStack().getAmount();
+    }
+  }
+
+  @EventHandler(priority = EventPriority.MONITOR)
+  public void onPlayerPickup(PlayerPickupItemEvent event) {
+    if (isTrackedEntity(event.getItem())) {
+      spawnedEntities = spawnedEntities - event.getItem().getItemStack().getAmount();
+    }
+  }
+
+  @EventHandler(priority = EventPriority.MONITOR)
+  public void onPlayerMove(CoarsePlayerMoveEvent event) {
+    Player player = event.getPlayer();
+    if (match.getPlayer(player).isObserving()) {
+      return;
+    }
+    if (definition.playerRegion.contains(event.getPlayer()) && trackedPlayers.get(player) == null) {
+      trackedPlayers.put(player, match.getPlayer(player));
+    } else if (!definition.playerRegion.contains(event.getPlayer())) {
+      trackedPlayers.remove(player);
+    }
+  }
+
+  @EventHandler(priority = EventPriority.MONITOR)
+  public void onPlayerQuit(PlayerQuitEvent event) {
+    trackedPlayers.remove(event.getPlayer());
+  }
```

Comment:
Since you're using `OnlinePlayerMapAdapter` players are automatically removed on quit, so you can remove this.